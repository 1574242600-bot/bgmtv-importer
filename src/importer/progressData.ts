import { CollectionStatusType, CollectionStatusTypeEnum } from '@bangumi/api/collection';
import { findDuplicate } from '@utils/index';

export class ProgressDataBox {
    private data: ProgressData[];
    constructor(data: ProgressData[]) {
        this.data = data;
    }

    getAll(): ProgressData[] {
        return this.data.slice();
    }
}

/** 
 * 合并重复的进度数据 (进度落后于其他id相同的数据)
 * 
 * */
export function mergeDuplicate(PD: ProgressDataBox) {
    const source = PD.getAll();
    
    const duplicate = findDuplicate(source, (v) => v.id + '');
    if (duplicate === null) return PD;

    const out: ProgressData[] = source
        .filter((v) => duplicate[v.id] === undefined); // 未重复项

    for (const id in duplicate) {
        const indexes = duplicate[id];
        const first = source[indexes.shift()];

        for (const index of indexes) {
            const pD = source[index];

            //优先级 dropped > collect > on_hold > do > wish
            if (pD.status) {
                const toValue = (status) => {
                    const map = {
                        [CollectionStatusTypeEnum.WISH]: 1,
                        [CollectionStatusTypeEnum.DO]: 2,
                        [CollectionStatusTypeEnum.ON_HOLD]: 3,
                        [CollectionStatusTypeEnum.COLLECT]: 4,
                        [CollectionStatusTypeEnum.DROPPED]: 5
                    };

                    return map[status];
                };

                first.status = toValue(pD.status) > toValue(first.status) 
                    ? pD.status 
                    : first.status;
            }

            if (pD.watchedEps) {
                pD.watchedEps.forEach((v) => {
                    if (!first.watchedEps.includes(v)) {
                        first.watchedEps.push(v);
                    }
                });
            }
            
            if (pD.watchedVols) {
                first.watchedVols = pD.watchedVols > first.watchedVols 
                    ? pD.watchedVols 
                    : first.watchedVols;
            }
        }
        
        out.push(first);
    }

    return new ProgressDataBox(out);
}

export function parsePDArr(data: ProgressData[]) {
    
    for (const index in data) {
        const pD = data[index];

        if (!isPD(pD)) {
            const text = `无效的 ProgressJsonData: 在位置 ${index} 检查到错误 \n`
                + `值: ${JSON.stringify(pD)}`;

            throw new TypeError(text);
        }
    }

    return new ProgressDataBox(data);
}

function isPD(pD: ProgressData) {
    function isValidStatus(status: string) {
        return status === CollectionStatusTypeEnum.WISH
            || status === CollectionStatusTypeEnum.COLLECT
            || status === CollectionStatusTypeEnum.DO
            || status === CollectionStatusTypeEnum.ON_HOLD
            || status === CollectionStatusTypeEnum.DROPPED;
    }




    const { id, status, watchedEps, watchedVols } = pD;

    if (!Number.isInteger(id) || id < 0) return false;

    if (typeof status === 'string' && !isValidStatus(status)) return false;

    if (watchedVols
        && (!Number.isInteger(watchedVols) || watchedVols <= 0)
    ) return false;

    if (watchedEps
        && (!(watchedEps instanceof Array)
                || !watchedEps.every((v) => v > 0)
        )
    ) return false;

    return true;
}


export interface ProgressData {
    id: number,
    status?: CollectionStatusType,
    watchedEps?: number[],
    watchedVols?: number
}