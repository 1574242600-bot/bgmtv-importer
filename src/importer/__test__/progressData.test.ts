import { parsePDArr, ProgressData, mergeDuplicate } from '../progressData';

test('parsePDArr(): 提供无效的参数, 抛出 type 异常', () => {

    const invalidId: ProgressData[] = [
        {
            id: -1,
        }
    ];

    const received = () => parsePDArr(invalidId);
    expect(() => received())
        .toThrow('在位置 0 检查到错误');


    const invalidStatus: ProgressData[] = [
        {
            id: 1,
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            /// @ts-ignore
            status: 'invalid',
        }
    ];

    const received1 = () => parsePDArr(invalidStatus);
    expect(() => received1())
        .toThrow('在位置 0 检查到错误');


    const invalidWatchedEps: ProgressData[] = [
        {
            id: 1,
            watchedEps: [0]
        }
    ];

    const received2 = () => parsePDArr(invalidWatchedEps);
    expect(() => received2())
        .toThrow('在位置 0 检查到错误');


    const invalidWatchedVols: ProgressData[] = [
        {
            id: 1,
            watchedVols: -1
        }
    ];

    const received3 = () => parsePDArr(invalidWatchedVols);
    expect(() => received3())
        .toThrow('在位置 0 检查到错误');
});

test('合并重复的进度数据', () => {
    const PDArr: ProgressData[] = [
        {
            id: 1,
            status: 'collect',
            watchedEps: [1, 2]
        },
        {
            id: 1,
            status: 'on_hold',
            watchedEps: [],
            watchedVols: 2
        },
        {
            id: 1,
            status: 'do',
            watchedEps: [1],
            watchedVols: 3
        },
        {
            id: 1,
            status: 'dropped',
            watchedEps: [1],
            watchedVols: 3
        },
        {
            id: 2,
            status: 'wish',
            watchedEps: [1, 2, 3],
            watchedVols: 2
        },
        {
            id: 2,
            status: 'do',
            watchedVols: 3
        },
        {
            id: 3,
            watchedVols: 3
        }
    ];

    const received = mergeDuplicate(parsePDArr(PDArr)).getAll();
    expect(received).toMatchSnapshot();
});