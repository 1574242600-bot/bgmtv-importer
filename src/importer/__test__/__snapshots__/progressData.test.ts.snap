// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`合并重复的进度数据 1`] = `
Array [
  Object {
    "id": 3,
    "watchedVols": 3,
  },
  Object {
    "id": 1,
    "status": "dropped",
    "watchedEps": Array [
      1,
      2,
    ],
    "watchedVols": undefined,
  },
  Object {
    "id": 2,
    "status": "do",
    "watchedEps": Array [
      1,
      2,
      3,
    ],
    "watchedVols": 3,
  },
]
`;
